plutus-core/executables/Common.hs

```diff
@@ -79,34 +86,44 @@ class Executable p where
 
   -- | Check a program for unique names.
   -- Throws a @UniqueError@ when not all names are unique.
-  checkProgram ::
+  checkUnique ::
      (Ord ann, AsUniqueError e ann,
        MonadError e m)
-    => (UniqueError ann -> Bool)
-    -> p ann
+    => p ann
     -> m ()
 
   -- | Convert names to de Bruijn indices and then serialise
   serialiseProgramFlat :: (Flat ann, PP.Pretty ann) => AstNameType -> p ann -> IO BSL.ByteString
 
-  -- | Read and deserialise a Flat-encoded UPLC AST
+  -- | Read and deserialise a Flat-encoded AST
   loadASTfromFlat :: AstNameType -> Input -> IO (p ())
 
--- | Instance for PLC program.
-instance Executable PlcProg where
-  parseNamedProgram inputName = PLC.runQuoteT . UPLC.parse PLC.program inputName
-  checkProgram = PLC.checkProgram
-  serialiseProgramFlat nameType p =
+-- For PIR and PLC
+serialiseTProgramFlat :: Flat a => AstNameType -> a -> IO BSL.ByteString
+serialiseTProgramFlat nameType p =
       case nameType of
         Named         -> pure $ BSL.fromStrict $ flat p
         DeBruijn      -> typedDeBruijnNotSupportedError
         NamedDeBruijn -> typedDeBruijnNotSupportedError
-  loadASTfromFlat = loadPlcASTfromFlat
+
+-- | Instance for PIR program.
+instance ProgramLike PirProg where
+  parseNamedProgram inputName = PLC.runQuoteT . PIR.parse PIR.program inputName
+  checkUnique _ = pure () -- decided that it's not worth implementing since it's checked in PLC.
```

Comment:
I think I'm generally in favor of checking as much as you can, but I don't mind it being arranged this way.

Reply:
yeah I wasn't sure about this either. It'll be quite a bit of (duplicated) code. I talked to Michael about this and came to this conclusion.